shared_model/backend/protobuf/commands/impl/proto_command.cpp

```diff
@@ -4,44 +4,92 @@
+  using ProtoCommandVariantType =
+      ::boost::variant<shared_model::proto::AddAssetQuantity,
+                       shared_model::proto::AddPeer,
+                       shared_model::proto::AddSignatory,
+                       shared_model::proto::AppendRole,
+                       shared_model::proto::CreateAccount,
+                       shared_model::proto::CreateAsset,
+                       shared_model::proto::CreateDomain,
+                       shared_model::proto::CreateRole,
+                       shared_model::proto::DetachRole,
+                       shared_model::proto::GrantPermission,
+                       shared_model::proto::RemoveSignatory,
+                       shared_model::proto::RevokePermission,
+                       shared_model::proto::SetAccountDetail,
+                       shared_model::proto::SetQuorum,
+                       shared_model::proto::SubtractAssetQuantity,
+                       shared_model::proto::TransferAsset>;
+
+  /// list of types in proto variant
+  using ProtoCommandListType = ProtoCommandVariantType::types;
+}  // namespace
 
 namespace shared_model {
   namespace proto {
 
-    template <typename CommandType>
-    Command::Command(CommandType &&command)
-        : CopyableProto(std::forward<CommandType>(command)),
-          variant_{[this] {
-            auto &&ar = *proto_;
-            int which = ar.GetDescriptor()
-                            ->FindFieldByNumber(ar.command_case())
-                            ->index();
-            return shared_model::detail::variant_impl<ProtoCommandListType>::
-                template load<ProtoCommandVariantType>(
-                    std::forward<decltype(ar)>(ar), which);
-          }},
-          ivariant_{detail::makeLazyInitializer(
-              [this] { return CommandVariantType(*variant_); })} {}
-
-    template Command::Command(Command::TransportType &);
-    template Command::Command(const Command::TransportType &);
-    template Command::Command(Command::TransportType &&);
-
-    Command::Command(const Command &o) : Command(o.proto_) {}
-
-    Command::Command(Command &&o) noexcept : Command(std::move(o.proto_)) {}
+    struct Command::Impl {
+      explicit Impl(TransportType &&ref) : proto_(std::move(ref)) {}
+      explicit Impl(const TransportType &ref) : proto_(ref) {}
+
+      detail::ReferenceHolder<TransportType> proto_;
+
+      ProtoCommandVariantType variant_{[this] {
+        auto &&ar = *proto_;
+        int which =
+            ar.GetDescriptor()->FindFieldByNumber(ar.command_case())->index();
+        return shared_model::detail::variant_impl<ProtoCommandListType>::
+            template load<ProtoCommandVariantType>(
+                std::forward<decltype(ar)>(ar), which);
+      }()};
+
+      CommandVariantType ivariant_{
+          [this] { return CommandVariantType(variant_); }()};
+    };
+
+    Command::Command(const Command &o) {
+      this->impl_ = std::make_unique<Impl>(*o.impl_->proto_);
```

Comment:
It should be possible to use delegating constructor: `: Command(*o.impl_->proto_)`